Adding new banners for BNPL to show upon legacy checkout deprecation

wjrosa opened the pull request:
<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

Fixes STRIPE-430
P2: pcUrMV-b0q-p2
Figma: nFI8jCCAQnd19PNwScqVU6-fi

## Changes proposed in this Pull Request:

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

We are introducing a new promotional banner surface to promote the Buy Now, Pay Later payment methods. This banner will show up for merchants using the new checkout experience that don't have BNPLs enabled, and after the legacy checkout is deprecated.

The banner should be dismissible. And should send merchants to `https://woocommerce.com/document/stripe/setup-and-configuration/additional-payment-methods/` upon clicking the "Learn more" link.

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

Preview:
<img width="765" alt="Screenshot 2025-05-29 at 18 13 42" src="https://github.com/user-attachments/assets/35454f81-c727-458d-915e-d87902d835ae" />

I am also taking the opportunity to improve how we organize the banner components' files.

## Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

- Checkout and build this branch on your test environment (`add/new-banners-for-bnpl-upon-legacy-checkout-deprecation`)
- As a merchant, enable the legacy checkout experience

### Rendering

- Confirm the banner is not shown yet
- Disable the legacy checkout experience
- Disable BNPL methods (Klarna, Affirm, Afterpay/Clearpay)
- Confirm the new banner is displayed in the payment methods page (`wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=methods`)
- Enable any BNPL method
- Confirm the banner is gone

### Visuals

- Confirm the visuals match the Figma file and the preview above

### Dismissing

- Click the "Dismiss" link
- Confirm the banner is gone

### Learn more link

- Click the "Learn more" link
- Confirm it sends you to: https://woocommerce.com/document/stripe/setup-and-configuration/additional-payment-methods/

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [ ] Tested on mobile (or does not apply)

### Changelog entry

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

wjrosa committed:
Adding new banners for BNPL to show upon legacy checkout deprecation

wjrosa committed:
Latest updates to the banner visuals

wjrosa committed:
Moving the banners to a new folder

wjrosa committed:
Better organizing files

wjrosa committed:
Banner style update

wjrosa committed:
Changing illustration paths

wjrosa committed:
Update learn more link

wjrosa committed:
Adding the condition required to display

wjrosa committed:
Unit tests

wjrosa committed:
Adding a legacy deprecation feature flag

wjrosa committed:
Unit tests

wjrosa committed:
Changelog and readme entries

wjrosa committed:
Checking the feature flag value

wjrosa committed:
Merge branch 'develop' into add/new-banners-for-bnpl-upon-legacy-checkout-deprecation

wjrosa committed:
Fix rebase

wjrosa committed:
Fix rebase

wjrosa committed:
Renaming the banner component

wjrosa committed:
Renaming the banner component

wjrosa committed:
Renaming the banner component

wjrosa committed:
Fix tests

wjrosa committed:
Changing the constant value to use the JS params

Mayisha reviewed (commented):
Thanks @wjrosa for the implementation. I am sharing a few UI observations.

- The `Learn more` button style looks different for me. The border is missing and it looks like a link only.

<img width="846" alt="Screenshot 2025-06-03 at 1 34 14 PM" src="https://github.com/user-attachments/assets/c79f1a6d-a16f-41e4-953c-43647c30ccf2" />

- When I enable Klarna/Affirm/AfterPay the banner is hidden as expected. But the banner appears briefly on page load.


https://github.com/user-attachments/assets/15d708fb-4d4e-4566-8cc2-c4838ab31701

- When I click the dismiss button, the banner disappears as expected. But it returns on page reload. Is it expected?

- The banner styles are off for mobile view.

<img width="496" alt="Screenshot 2025-06-03 at 1 53 21 PM" src="https://github.com/user-attachments/assets/c33b2979-59bf-465a-ba22-f26da7ef6f9a" />

Mayisha commented on client/settings/payment-settings/promotional-banner/bnpl-promotion-banner.js:
```diff
@@ -0,0 +1,88 @@
+						<br />
+						{ __(
+							'Affirm and Klarna payments are auto-enabled with Stripe for eligible merchants.',
+							'woocommerce-gateway-stripe'
```
QQ: is AfterPay not part of the promotion?

wjrosa replied:
I don't think so. But double-checking in pcUrMV-b0q-p2#comment-32116

wjrosa replied:
@Mayisha confirmed that Afterpay is not part of the promotion (see pcUrMV-b0q-p2#comment-32318)

Mayisha commented on client/settings/payment-settings/promotional-banner/index.js:
```diff
@@ -0,0 +1,84 @@
+	oauthUrl,
+	testOauthUrl,
+} ) => {
+	const isLegacyDeprecated =
```
```suggestion
	const isUpeEnabled =
```

Nitpick

wjrosa committed:
Update client/settings/payment-settings/promotional-banner/index.js

Co-authored-by: Mayisha <33387139+Mayisha@users.noreply.github.com>

wjrosa committed:
Removing unnecessary variable

wjrosa committed:
Adding new setting in the backend to control the BNPL banner visibility

wjrosa committed:
BNPL banner constant identifier

wjrosa committed:
Handling display of the banner using the new backend setting

wjrosa committed:
New request on the frontend to disable the banner visibility

wjrosa committed:
Updating the banner type with the new type

wjrosa committed:
Fix tests

wjrosa committed:
Fix tests

wjrosa committed:
Fix banner style for mobile

wjrosa commented:
Thanks for the great review, @Mayisha !

> The Learn more button style looks different for me. The border is missing and it looks like a link only.

That's because we use a different component for external links. I asked if that's OK in pcUrMV-b0q-p2#comment-32116

> When I enable Klarna/Affirm/AfterPay the banner is hidden as expected. But the banner appears briefly on page load.

Still looking 👀 

> When I click the dismiss button, the banner disappears as expected. But it returns on page reload. Is it expected?

It was for the other banners, but I think this one should have a way to be permanently dismissed. I added a backend setting in my latest commits, updated the notices endpoint, and handled the display on the frontend using it.

> The banner styles are off for mobile view.

Good catch! Fixed it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4328/commits/bbf8a4d5e40402c9cff3725de9d4ff2df198a715

wjrosa committed:
Merge branch 'develop' into add/new-banners-for-bnpl-upon-legacy-checkout-deprecation

wjrosa committed:
Fix banner quickly displaying when it should be hidden

wjrosa committed:
Reloading page upon dismissing banner

wjrosa committed:
Fix tests

wjrosa commented on includes/admin/class-wc-stripe-settings-controller.php:
```diff
@@ -231,26 +231,32 @@ public function admin_scripts( $hook_suffix ) {
+			'stripe_oauth_url'             => $oauth_url,
+			'stripe_test_oauth_url'        => $test_oauth_url,
+			'show_customization_notice'    => get_option( 'wc_stripe_show_customization_notice', 'yes' ) === 'yes' ? true : false,
+			'show_bnpl_promotional_banner' => $show_bnpl_promotion_banner,
```
That's the only new line, but I had to fix the CS alignment issue.

wjrosa commented on includes/admin/class-wc-rest-stripe-settings-controller.php:
```diff
@@ -740,13 +740,33 @@ public function update_individual_payment_method_settings( WP_REST_Request $requ
+	 * @param WP_REST_Request $request Request object.
+	 * @return WP_REST_Response
+	 */
+	public function dismiss_notice( WP_REST_Request $request ) {
```
I made this method more generic, instead of creating another specific one for the `wc_stripe_show_bnpl_promotion_banner` setting.

wjrosa commented on client/settings/payment-settings/promotional-banner/re-connect-account-banner.js:
```diff
@@ -0,0 +1,79 @@
+import { __ } from '@wordpress/i18n';
```
I just moved this component to a new file.

wjrosa commented on client/settings/payment-settings/promotional-banner/new-checkout-experience-banner.js:
```diff
@@ -0,0 +1,111 @@
+import { __ } from '@wordpress/i18n';
```
I just moved this component to a new file.

wjrosa commented on client/settings/payment-settings/promotional-banner/new-checkout-experience-apms-banner.js:
```diff
@@ -0,0 +1,132 @@
+/* global wc_stripe_settings_params */
```
I just moved this component to a new file.

wjrosa commented on client/settings/payment-settings/promotional-banner/index.js:
```diff
@@ -0,0 +1,65 @@
+	testOauthUrl,
+} ) => {
+	let BannerContent = null;
+	switch ( promotionalBannerType ) {
```
Improved and simplified how we decide which banner to show.

wjrosa commented on client/settings/payment-settings/promotional-banner/get-promotional-banner-type.js:
```diff
@@ -0,0 +1,47 @@
+ * @param {Array} enabledPaymentMethodIds List of enabled payment method IDs.
+ * @return {null|string} The type of promotional banner to display, or null if no banner is applicable.
+ */
+export const getPromotionalBannerType = (
```
Moved the logic to decide the banner to show here.

wjrosa commented on client/settings/payment-settings/promotional-banner/banner-layout.js:
```diff
@@ -0,0 +1,47 @@
+import styled from '@emotion/styled';
```
This file now contains all the shared banner styled code.

wjrosa commented:
> When I enable Klarna/Affirm/AfterPay the banner is hidden as expected. But the banner appears briefly on page load.

Should be fixed with my latest commits 👍 . I had to do a refactoring, making the rendering a bit less async (getting the initial value from the backend for this specific banner).

Mayisha committed:
Merge branch 'develop' into add/new-banners-for-bnpl-upon-legacy-checkout-deprecation

Mayisha commented on includes/admin/class-wc-rest-stripe-settings-controller.php:
```diff
@@ -740,13 +740,33 @@ public function update_individual_payment_method_settings( WP_REST_Request $requ
 	 *
 	 * @return WP_REST_Response
+	 *
+	 * @deprecated since 9.6.0, use `dismiss_message` instead.
```
```suggestion
	 * @deprecated since 9.6.0, use `dismiss_notice` instead.
```

Mayisha reviewed (approved):
Thanks for the fixes @wjrosa. The UI issues are all fixed now. 👍

wjrosa committed:
Update includes/admin/class-wc-rest-stripe-settings-controller.php

Co-authored-by: Mayisha <33387139+Mayisha@users.noreply.github.com>
